crates/nu-parser/src/parser.rs

```diff
@@ -2034,6 +2034,10 @@ pub fn parse_brace_expr(
         } else {
             parse_record(working_set, span)
         }
+    } else if matches!(shape, SyntaxShape::Block) {
+        parse_block_expression(working_set, span)
+    } else if matches!(shape, SyntaxShape::MatchBlock) {
+        parse_match_block_expression(working_set, span)
     } else if matches!(second_token_contents, Some(TokenContents::Pipe))
         || matches!(second_token_contents, Some(TokenContents::PipePipe))
     {
```

Comment:
Instead of priorize the SyntaxShape checking for block and MatchBlock, I think it's better to check SyntaxShape before parsing as a closure:
```suggestion
        // when meeting Pipe or PipePipe, it's certain to be a closure
        // But nushell should guide the checking process though SyntaxShape
        if matches!(shape, SyntaxShape::Block) {
            working_set.error(ParseError::ExpectedWithStringMsg(
                "expect a block, but finding a closure".to_string(),
                span,
            ));
            return Expression::garbage(working_set, span);
        }

```

Reply:
Thanks! I've applied your feedback with a few changes